src/mlpack/core/metrics/lmetric_impl.hpp

```diff
@@ -61,7 +61,19 @@ typename VecTypeA::elem_type LMetric<2, true>::Evaluate(
     const VecTypeA& a,
     const VecTypeB& b)
 {
-  return arma::norm(a - b, 2);
+  if ((std::is_same<typename VecTypeA::elem_type, double>::value or
+     std::is_same<typename VecTypeA::elem_type, float>::value) and
+     (std::is_same<typename VecTypeB::elem_type, double>::value or
+     std::is_same<typename VecTypeB::elem_type, float>::value))
+  {
+    return arma::norm(a - b, 2);
```

Comment:
This will still fail to compile if `VecTypeA` or `VecTypeB` contain integer elements, as `arma::norm()` works only for vectors/matrices with floating point values. Whenever a template is instantiated, all code paths must be valid, including the path for `if(false) { ... }`. This is a (current) limitation of  C++.

To fix this, either forcefully convert the `a` and `b` inputs (regardless of their type) via `arma::conv_to` to vectors/matrices with floating point values before calling `arma::norm()`, or use two separate functions (one with conversion and one without conversion) with the instantiation of each function controlled via SFINAE restrictors.  The former approach is less efficient than the latter approach.

Reply:
I need to check on both of the solutions. SFINAE would not be great here since there are already several overloads for the `Evaluate` functions already. If I am going to add SFINAE there will be at least 12 `Evaluate` functions that are not funny to maintain.

Reply:
Only SFINAE wrappers around `arma::norm()` would be required.  Say the resulting two functions would be called `mlpack_norm(A,B)` (one for vectors with floating point elements, one for other element types).  Then within `Evaluate()`, all calls to `arma::norm(A,B)` would be replaced with `mlpack_norm(A,B)`.

Reply:
Right, it should be possible to do this with just two overloads.  You can use a complex conditional for SFINAE, and you could even use things like `arma::is_supported_blas_type<>`, although technically that is not a part of the public API so it is potentially subject to change in newer Armadillo versions; it may therefore be better to make an equivalent struct in the mlpack codebase.